### PR TITLE
A class to represent polynomials in the monomial basis

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

File: benchmarks/polynomial.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+using quantities::Length;
+using quantities::Quantity;
+using quantities::SIUnit;
+using quantities::Time;
+
+namespace numerics {
+
+namespace {
+int const evaluations_per_iteration = 1000;
```
Comment (eggrobin) [inline benchmarks/polynomial.cpp]: constexpr int

File: benchmarks/polynomial.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      break;
+    case 16:
+      EvaluatePolynomialInMonomialBasis<double, Time, 16>(state);
+      break;
+    default:
+      LOG(FATAL) << "Degree " << degree << " in BM_EvaluatePolynomialDouble";
+  }
+}
+
+void BM_EvaluatePolynomialInMonomialBasisQuantity(benchmark::State& state) {
```
Comment (eggrobin) [inline benchmarks/polynomial.cpp]: Add a benchmark with vectors, and one with points (possibly one with degrees of freedom too).

File: numerics/polynomial.hpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+//       Value,
+//       Quotient<Difference<Value>,
+//                Exponentiation<Difference<Argument>, order>>>;
+//
+//   template<typename Value, typename Argument>
+//   using Derivative = NthDerivative<Value, Argument, 1>;
+//
+// Unfortunately VS2015 is buggy and this interacts poorly with the
+// std::integer_sequence below (we get the wrong types).  Revisit once MSFT has
+// fixed their bugs.
```
Comment (eggrobin) [inline numerics/polynomial.hpp]: I'm not sure this is going to happen unless we make a bug report, considering that https://developercommunity.visualstudio.com/content/problem/106570/incorrect-compilation-error-on-c-code-involving-te.html only kicked the bug down the road to https://developercommunity.visualstudio.com/content/problem/165535/incorrect-compilation-error-on-c-operators-and-dec.html.

Maybe we should consider switching to another compiler front-end...

File: quantities/quantities_body.hpp [hunk 4]
```diff
@@ -298,13 +292,45 @@ constexpr bool Quantity<D>::operator!=(Quantity const& right) const {
...
+                "Invalid angle exponent");
   message->set_dimensions(D::representation);
   message->set_magnitude(magnitude_);
 }
 
 template<typename D>
 Quantity<D> Quantity<D>::ReadFromMessage(
     serialization::Quantity const& message) {
+  static_assert(D::length_is_serializable,
+                "Invalid length exponent");
```
Comment (eggrobin) [inline quantities/quantities_body.hpp]: Should we keep this stuff in the message? it makes the messages larger, adds quite a bit of ugly code, and I have never seen it fail (we never change the dimensions of a field, we add new fields instead).

File: benchmarks/polynomial.cpp (hunk 1 above)
Comment (pleroy) [reply benchmarks/polynomial.cpp]: Done.

File: benchmarks/polynomial.cpp (hunk 2 above)
Comment (pleroy) [reply benchmarks/polynomial.cpp]: This will be in a future PR.

File: numerics/polynomial.hpp (hunk 3 above)
Comment (pleroy) [reply numerics/polynomial.hpp]: I guess I should create a bug with MSFT.

File: quantities/quantities_body.hpp (hunk 4 above)
Comment (pleroy) [reply quantities/quantities_body.hpp]: Let's leave it that way for now.

**Pull request merged**